### PR TITLE
Fix and enable a large subset of tests on the z/OS OMR CI

### PR DESCRIPTION
**Pull request by fjeremic**
In this PR as a sequence of commits we fix a whole host of issues to get _most_ of the tests currently ran by `ctest` working on z/OS. We can use `ctest` on z/OS as well, however at the moment the thread extended test is hanging and is preventing us from making the switch. Several other tests have failures which need to be investigated. However we are well on our way.

The following table represents the list of tests currently running on other platforms and the before/after work once this PR is merged. 

| Test name             | Runs before this PR | Runs after this PR | Test status        | Reason/Comment        |
|-----------------------|---------------------|--------------------|--------------------|-----------------------|
| gcexample             |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| python "run_tests.py" |                     | :heavy_check_mark: | :heavy_check_mark: |  |
| conditionals          |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| isSupportedType       |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| iterfib               |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| nestedloop            |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| pow2                  |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| simple                |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| worklist              |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| omralgotest           | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |                       |
| TestBytes             |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| TestIntrusiveList     |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| TestTypeTraits        |                     | :heavy_check_mark: | :heavy_check_mark: |                       |
| omrporttest           |                     | :x:                | :x:                | #4747 (excluded via ctest)          |
| omrrastest            | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |                       |
| omrsubscribertest     | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |                       |
| omrtraceoptiontest    | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |                       |
| omrsigtest            | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |                       |
| omrthreadextendedtest |                     | :x:                | :x:                | #4746 (excluded via ctest)                  |
| omrthreadtest 1       |                     | :heavy_check_mark: | :heavy_check_mark:      | Pthread issues?        |
| omrthreadtest 2       |                     | :heavy_check_mark: | :heavy_check_mark:      | Pthread issues?        |
| omrthreadtest 3       |                     | :heavy_check_mark: | :heavy_check_mark:      | Pthread issues?        |
| omrutiltest           | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |                       |
| omrvmtest             | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |                       |
| omrgctest             |                     | :heavy_check_mark:                | :heavy_check_mark:                |                |
| jitbuildertest        |                     | :heavy_check_mark:                | :heavy_check_mark:                |       |
| compilertest          |                     | :x:                | :x:                | #4719 (excluded via ctest)               |
| triltest              |                     | :heavy_check_mark:                | :heavy_check_mark:                |               |
| comptest              |                     | :heavy_check_mark: | :heavy_check_mark: |                       |

~The end goal will be to get all tests functionally running on z/OS and have 100% parity with other platforms running via `ctest`.~ Goal has been achieved. We are currently running all checkmarked tests above via `ctest` on z/OS!

### CODE REVIEW

**Comment by fjeremic** [comment]
@genie-omr build zos

File: jitbuilder/control/Jit.cpp [hunk 1]
```diff
@@ -229,4 +229,6 @@ internal_shutdownJit()
 
    TR::CodeCacheManager &codeCacheManager = fe->codeCacheManager();
    codeCacheManager.destroy();
+
+   TR::CompilationController::shutdown();
```
Comment (fjeremic) [inline jitbuilder/control/Jit.cpp]: This was a pretty scary one. We are likely not cleaning up a ton of other resources. We really ought to take a hard look at what a proper shutodwn sequence should look like and implement it at a common level. Currently we have a ton of duplicated logic between JitBuilder and TestCompiler for example.

**Comment by fjeremic** [comment]
Suggesting @mstoodle and @Leonardo2718 for reviews.

**Comment by fjeremic** [comment]
@genie-omr build all

**Comment by fjeremic** [comment]
Latest commit enables `triltest` to run as well.

File: compiler/ilgen/OMRMethodBuilder.hpp [hunk 2]
```diff
@@ -351,7 +351,7 @@ class MethodBuilder : public TR::IlBuilder
    TR::IlType                ** _cachedParameterTypes;
    const char                * _definingFile;
    char                        _definingLine[MAX_LINE_NUM_LEN];
-   TR::IlType                * _cachedParameterTypesArray[10];
+   TR::IlType                * _cachedParameterTypesArray[18];
```
Comment (fjeremic) [inline compiler/ilgen/OMRMethodBuilder.hpp]: This seems less than ideal. There should be a better way of doing this than hardcoding a constant array length and asserting that the number of parameters is less than 18. We should use a dynamic data structure and allow as many parameters as the user wants (within reason).

**Comment by fjeremic** [comment]
@genie-omr build all

File: buildenv/jenkins/jobs/builds/Build-zos_390-64 [hunk 3]
```diff
@@ -42,7 +42,7 @@ pipeline {
                 timestamps {
                     dir('build') {
                         echo "Sanity Test..."
-                        sh'''./omralgotest -avltest:fvtest/algotest/avltest.lst && ./omrvmtest && ./omrutiltest && ./omrsigtest && ./omrrastest && ./omrsubscribertest && ./omrtraceoptiontest'''
+                        sh'''ctest -V'''
```
Comment (mstoodle) [inline buildenv/jenkins/jobs/builds/Build-zos_390-64]: not your mistake, but why doesn't this file have a copyright header?

File: buildenv/jenkins/jobs/pull-requests/PullRequest-zos_390-64 [hunk 4]
```diff
@@ -28,7 +28,7 @@ pipeline {
                 timestamps {
                     dir('build') {
                         echo "Sanity Test..."
-                        sh'''./omralgotest -avltest:fvtest/algotest/avltest.lst && ./omrvmtest && ./omrutiltest && ./omrsigtest && ./omrrastest && ./omrsubscribertest && ./omrtraceoptiontest'''
+                        sh'''ctest -V'''
```
Comment (mstoodle) [inline buildenv/jenkins/jobs/pull-requests/PullRequest-zos_390-64]: same problem here: no copyright header

File: compiler/ilgen/OMRMethodBuilder.hpp (hunk 2 above)
Comment (mstoodle) [reply compiler/ilgen/OMRMethodBuilder.hpp]: Agreed, but you shouldn't have to fix that here. Created #4748 .

File: jitbuilder/control/Jit.cpp (hunk 1 above)
Comment (mstoodle) [reply jitbuilder/control/Jit.cpp]: Agree. Opened #4749 .

File: jitbuilder/release/cpp/samples/RecursiveFib.cpp [hunk 5]
```diff
@@ -101,6 +96,11 @@ RecursiveFibonnaciMethod::buildIL()
    recursiveCase->         Sub(
    recursiveCase->            Load("n"),
    recursiveCase->            ConstInt32(2)))));
+   
+#if defined(RFIB_DEBUG_OUTPUT)
```
Comment (mstoodle) [inline jitbuilder/release/cpp/samples/RecursiveFib.cpp]: weird...good spot.

File: buildenv/jenkins/jobs/builds/Build-zos_390-64 (hunk 3 above)
Comment (mstoodle) [reply buildenv/jenkins/jobs/builds/Build-zos_390-64]: Opened #4753 .

File: buildenv/jenkins/jobs/pull-requests/PullRequest-zos_390-64 (hunk 4 above)
Comment (mstoodle) [reply buildenv/jenkins/jobs/pull-requests/PullRequest-zos_390-64]: Opened #4753 .

**Comment by fjeremic** [comment]
@genie-omr build zos

File: fvtest/jitbuildertest/SystemLinkageTest.cpp [hunk 6]
```diff
@@ -53,9 +69,19 @@ class SystemLinkageTest : public JitBuilderTest {};
...
    {
+#if defined(PPC)
+   // TODO (#4765): Fix the Power linkage and enable this test
+   // TODO (#4764): This is a poor man's SKIP_IF which is only available for Tril tests currently. We should find a
+   // nice home for this useful utility so we can use it in JitBuilder tests as well.
+   const ::testing::TestInfo* const test_info = ::testing::UnitTest::GetInstance()->current_test_info();
+   ::testing::Test::RecordProperty("skipped", "Known Bug");
+   std::cout << "Known Bug" << ": Skipping test: " << test_info->name() << "\n    " << "Power system linkage cannot handle stack arguments (see issue #4765)" << "\n";
+   SUCCEED() << "Power system linkage cannot handle stack arguments (see issue #4765)";
+#endif
```
Comment (fjeremic) [inline fvtest/jitbuildertest/SystemLinkageTest.cpp]: This is another hack we are adding in, but I don't think this PR is the correct place to solve this. I've opened up two new issues to address the hack added here.

**Comment by fjeremic** [comment]
This issue seems to be turning into a goldmine of spawning other issues. To avoid scope creep I will stop here as the number of fixes is getting large now. The task at hand has been achieved. We have parity now on z/OS and are running almost all the tests via `ctest` :yay:

One final whirl:

@genie-omr build all

**Comment by fjeremic** [comment]
This issue seems to be turning into a goldmine of spawning other issues. To avoid scope creep I will stop here as the number of fixes is getting large now. The task at hand has been achieved. We have parity now on z/OS and are running almost all the tests via `ctest` :yay:

One final whirl:

@genie-omr build all

File: compiler/z/codegen/SystemLinkagezOS.cpp [hunk 7]
```diff
@@ -282,6 +282,91 @@ void TR::S390zOSSystemLinkage::createPrologue(TR::Instruction* cursor)
...
+         case TR::VectorInt32:
+         case TR::VectorInt64:
+         case TR::VectorDouble:
+            {
+            if (numVRFArgs < self()->getNumVectorArgumentRegisters())
+               {
+               lri = numVRFArgs;
+               }
+
+            numVRFArgs++;
```
Comment (joransiu) [inline compiler/z/codegen/SystemLinkagezOS.cpp]: Reading the Vendor Interfaces, I think the same change for FPRs above to bump numGPRArgs++ should be applicable for VRs as well.

Comment (fjeremic) [reply compiler/z/codegen/SystemLinkagezOS.cpp]: I couldn't find an updated reference for XPLINK with vector registers. Do they even participate in linkage conventions? How does one pass a vector register argument? If you have any links that would be appreciated. I'll make the change though.

Comment (joransiu) [reply compiler/z/codegen/SystemLinkagezOS.cpp]: I was referencing: https://www-01.ibm.com/servers/resourcelink/svc00100.nsf/pages/zOSV2R3SA380688/$file/ceev100_v2r3.pdf around page 143/144, and the examples around page 911.

Comment (fjeremic) [reply compiler/z/codegen/SystemLinkagezOS.cpp]: Fixed in a03a579

**Comment by fjeremic** [comment]
@mstoodle do you want to be a committer for this one?

@genie-omr build all

**Comment by fjeremic** [comment]
@genie-omr build arm

**Comment by fjeremic** [comment]
@genie-omr build aarch64

**Comment by mstoodle** [comment]
Nice to see all those tests running and passing on z/OS ! Even some JitBuilder tests are now enabled and passing! Woohoo!

Once the appveyor tests run and pass, just for pedantry, I'll merge :) .

**Comment by mstoodle** [comment]
Thanks @fjeremic!

File: compiler/z/codegen/SystemLinkagezOS.cpp (hunk 7 above)
Comment (zl-wang) [reply compiler/z/codegen/SystemLinkagezOS.cpp]: Ever consider the alignment requirement of passing Vector arguments, i.e.  leaving more holes in GPR side? or, is there alignment requirement at all?  I remembered I did all these on AIX side before, need to read the code in details to see why it failed now.

**Pull request merged**